proto/usr/triton/bin/register-agent

```diff
@@ -0,0 +1,261 @@
+    local service_uuid
+    local retry=10
+    local url
+    local data
+
+    if [[ -z "${instance_uuid}" ]]; then
+        fatal 'must pass instance_uuid'
+    fi
+
+    while (( retry-- > 0 )); do
+        #
+        # Fetch the UUID of the SAPI service for this agent.
+        #
+        url="${SAPI_URL}/services?type=agent&name=${AGENT}"
+        if ! service_uuid="$(curl -sSf -H 'Accept: application/json' "${url}" \
+          | json -Ha uuid)"; then
+            printf 'Could not retrieve SAPI service UUID for "%s"\n' \
+              "${AGENT}" >&2
+            sleep 5
+            continue
+        fi
+
+        #
+        # Attempt to register the SAPI instance for this agent installation.
+        # We need not be overly clever here; SAPI returns success for a
+        # duplicate agent adoption.
+        #
+        url="${SAPI_URL}/instances"
+        data="{
+            \"service_uuid\": \"${service_uuid}\",
+            \"uuid\": \"${instance_uuid}\"
+        }"
+        if ! curl -sSf -X POST -H 'Content-Type: application/json' \
+          -d "${data}" "${url}"; then
+            printf 'Could not register SAPI instance with UUID "%s"\n' \
+              "${instance_uuid}" >&2
+            sleep 5
+            continue
+        fi
+
+        printf 'Agent successfully adopted into SAPI.\n' >&2
+        return 0
+    done
+
+    fatal 'adopt_instance: failing after too many retries'
+}
+
+#
+# The "config-agent" service reads configuration from JSON-formatted files in a
+# well-known local directory.  These configuration files tell "config-agent"
+# where to find local SAPI manifests describing the configuration for this
+# agent.
+#
+function add_config_agent_instance
+{
+    local instance_uuid="${1}"
+    local config_etc_dir="${ETC_DIR}/config-agent.d"
+    local agent_json="${config_etc_dir}/${AGENT}.json"
+    local data
+
+    if [[ -z "${instance_uuid}" ]]; then
+        fatal 'must pass in instance_uuid'
+    fi
+
+    mkdir -p "$config_etc_dir"
+
+    data="{
+        \"instance\": \"${instance_uuid}\",
+        \"localManifestDirs\": [
+            \"${AGENT_DIR}\"
+        ]
+    }"
+    if ! printf '%s' "${data}" | json >"${agent_json}"; then
+        fatal 'could not write configuration for "config-agent" (%s)' \
+          "${agent_json}"
+    fi
+
+    return 0
+}
+
+#
+# If there is an installed, running, instance of "config-agent", then restart
+# it now.  This ensures that config-agent will notice the addition of any local
+# manifests that we just installed.
+#
+function config_agent_restart
+{
+    if [[ "$(/usr/bin/systemctl is-active triton-config-agent)" == "active" ]]; then
+        /usr/bin/systemctl reload-or-restart triton-config-agent
+    fi
+
+    return 0
+}
+
+#
+# Check if we expect SAPI to be available.  Generally, registering with SAPI is
+# a hard requirement for the correct functioning of the system, but this
+# postinstall script can also be run during headnode setup; SAPI is not yet
```

Comment:
Whoa wait... headnode setup?  More copied-over-comments?

Reply:
Yeah, this script is ported from the agents `postinstall.sh` script. Eventually (especially with agent stuff) I want to refactor with SmartOS stuff to reduce code duplication.